src/webgpu/shader/execution/expression/case_cache.ts

```diff
@@ -0,0 +1,202 @@
+ * This form can be safely encoded to JSON.
+ */
+export type SerializedCase = {
+  input: SerializedValue | SerializedValue[];
+  expected: SerializedExpectation;
+};
+
+/** serializeCase() converts an Case to a SerializedCase */
+export function serializeCase(c: Case): SerializedCase {
+  return {
+    input: c.input instanceof Array ? c.input.map(v => serializeValue(v)) : serializeValue(c.input),
+    expected: serializeExpectation(c.expected),
+  };
+}
+
+/** serializeCase() converts an SerializedCase to a Case */
+export function deserializeCase(data: SerializedCase): Case {
+  return {
+    input:
+      data.input instanceof Array
+        ? data.input.map(v => deserializeValue(v))
+        : deserializeValue(data.input),
+    expected: deserializeExpectation(data.expected),
+  };
+}
+
+/** CaseListBuilder is a function that builds a CaseList */
+export type CaseListBuilder = () => CaseList;
+
+/**
+ * CaseCache is a cache of CaseList.
+ * CaseCache implements the Cacheable interface, so the cases can be pre-built
+ * and stored in the data cache, reducing computation costs at CTS runtime.
+ */
+export class CaseCache implements Cacheable<Record<string, CaseList>> {
+  /**
+   * Constructor
+   * @param name the name of the cache. This must be globally unique.
+   * @param builders a Record of case-list name to case-list builder.
+   */
+  constructor(name: string, builders: Record<string, CaseListBuilder>) {
+    this.path = `webgpu/shader/execution/case-cache/${name}.json`;
+    this.builders = builders;
+  }
+
+  /** get() returns the list of cases with the given name */
+  public async get(name: string): Promise<CaseList> {
+    const data = await dataCache.fetch(this);
+    return data[name];
+  }
+
+  /**
+   * build() implements the Cacheable.build interface.
+   * @returns the data.
+   */
+  build(): Promise<Record<string, CaseList>> {
+    const built: Record<string, CaseList> = {};
+    for (const name in this.builders) {
+      const cases = this.builders[name]();
+      built[name] = cases;
+    }
+    return Promise.resolve(built);
+  }
+
+  /**
+   * serialize() implements the Cacheable.serialize interface.
+   * @returns the serialized data.
+   */
+  serialize(data: Record<string, CaseList>): string {
+    const serialized: Record<string, SerializedCase[]> = {};
+    for (const name in data) {
+      serialized[name] = data[name].map(c => serializeCase(c));
+    }
+    return JSON.stringify(serialized);
+  }
+
+  /**
+   * deserialize() implements the Cacheable.deserialize interface.
+   * @returns the deserialize data.
+   */
+  deserialize(serialized: string): Record<string, CaseList> {
+    const data = JSON.parse(serialized) as Record<string, SerializedCase[]>;
+    const casesByName: Record<string, CaseList> = {};
+    for (const name in data) {
+      const cases = data[name].map(caseData => deserializeCase(caseData));
+      casesByName[name] = cases;
+    }
+    return casesByName;
+  }
```

Comment:
You may be able to simplify/optimize some of this with the following:
- If `JSON.stringify` tries to serialize something with a `toJSON` method, it will call that and serialize the result
- `JSON.stringify` has `replacer` and `JSON.parse` has `reviver` that give you a chance to customize the JSON representation of a thing

Reply:
ack